### DISCUSSION TITLE
Deleting old application and appproj resources

### DISCUSSION THREAD

**Discussion by noam-codefresh** [argoproj/argo-cd#5393] [Q&A]
I am playing around with argo-cd on my cluster, and I keep adding and removing the entire installation (sometimes by using `kubectl delete...`, sometimes by deleting the argocd namespace). I am also doing some non-conventional things (like having argo-cd manage itself by first bootstrapping it, and then pointing it at an application that manages itself). This is probably causing some of my issues.

I notice that many times I have leftover resources in my cluster. When I delete the argocd namespace, the deletion sometimes hangs before completion. I guess the controller is already gone, when there are still "application" and "appproj" resources waiting for the finalizer to clean them up.

I found a tip about manually deleting the CRD definitions - they also usually hang, but then I edit them, and manually remove their finalizers. Then the CRDs finally delete themselves, and the argocd namespace is gone.

But I now realized that on my next argocd install, I get the CRDs again (naturally), and then the old "application" and "appproj" resources re-appear in my cluster (or at least - I can see them again).

So how can I delete such orphan resources? Is there some known workaround for this problem?

**Comment by noam-codefresh** [comment]
I ended up using the workaround from [here](https://medium.com/@craignewtondev/how-to-fix-kubernetes-namespace-deleting-stuck-in-terminating-state-5ed75792647e).

The tricky bit was to re-create the namespaces I deleted before (when I manually removed the finalizers from the CRDs themselves), and then I executed the `patch --raw` command on each of those orphaned resources (after manually removing the finalizers from their JSON output).

I hope this doesn't leave any more hidden leftovers in my cluster.

**Reply by alexmt** [reply]
Hello @noam-codefresh ,

Did you add `resources-finalizer.argocd.argoproj.io` finalizer to Argo CD application or project? In this case application controller cannot be deleted before all Application/AppProject CRs are deleted.

**Reply by noam-codefresh** [reply]
I read yesterday about adding it, to allow delete cascading. But it seems as though in my current setup, it is already there. So the default behavior when deleting the k8s resource would cause cascading. And if I want to delete without cascading - I should probably remove it from the App resource.

**Reply by dhess** [reply]
I'm doing similar things as @noam-codefresh and having the exact same issues. I've tried with and without `resources-finalizer.argocd.argoproj.io`, but the result is the same &mdash; `kubectl delete namespace argocd` hangs forever and if I try to clean things up by hand, like @noam-codefresh I have leftover application resources from the previous install when I create the new installation.

@noam-codefresh Did you ever get to the bottom of this? At this point I've concluded that if I want to use Argo CD, I'll probably need to create a cluster just for it so that I can wipe the cluster and recreate it from scratch if I have to.

**Reply by noam-codefresh** [reply]
I figured out the problem in my scenario was that argo-cd was managing itself, *and* it had the finalizer set up. So when I tried deleting the argo-cd *Application* (which handles itself), it would delete the actual deployments/pods (`argocd-server`, etc.) that were supposed to perform the finalization on itself (and on other applications).
The solution for me was just making sure the argo-cd *Application* has no finalizer. So now I can just delete all the applications, and nothing is actually deleting argo-cd deployments and pods. After all the applications are gone, I can safely remove the entire namespace to clean up the left-over deployments.

**Reply by cooervo** [reply]
>  Did you add `resources-finalizer.argocd.argoproj.io` finalizer to Argo CD application or project? In this case application controller cannot be deleted before all Application/AppProject CRs are deleted.

this is what I needed example adding :

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: my-application
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
```

now when I delete app, related resources are deleted as well

**Comment by mahmud2011** [comment]
```zsh
kubectl patch app APPNAME  -p '{"metadata": {"finalizers": ["resources-finalizer.argocd.argoproj.io"]}}' --type merge
kubectl delete app APPNAME
```

Does cascade also delete the `appproject`?
